testsuite/cluster-test/src/health/fullnode_check.rs

```diff
@@ -0,0 +1,125 @@
+                    &start,
+                    &end,
+                    10, /* step */
+                )
+                .map_err(|e| format_err!("No sync committed version data: {}", e))
+        })
+    }
+}
+
+pub fn convert_index(input: HashMap<String, f64>) -> HashMap<String, Vec<i64>> {
+    let mut res = HashMap::new();
+    for (peer_name, version) in input {
+        if let Some(cap) = VAL_REGEX.captures(peer_name.as_ref()) {
+            // this is index of validator
+            if let Some(cap) = cap.get(1) {
+                let index = cap.as_str();
+                let mut versions = vec![];
+                versions.push(-version as i64);
+                // we record version of validator to be a negative number
+                // to calculate diff with all full nodes are associated
+                res.insert(index.to_string(), versions);
+            }
+        } else if let Some(cap) = FULLNODE_REGEX.captures(peer_name.as_ref()) {
+            // this is index of fullnode
+            if let Some(cap) = cap.get(1) {
+                let index = cap.as_str();
+                if res.contains_key(index) {
+                    res.get_mut(index).unwrap().push(version as i64);
+                } else {
+                    let mut versions = vec![];
+                    versions.push(version as i64);
+                    res.insert(index.to_string(), versions);
+                }
+            }
+        } else {
+            panic!(
+                "Failed to parse peer name {} into validator_index",
+                peer_name
+            )
+        }
+    }
+    res
+}
+
+impl HealthCheck for FullNodeHealthCheck {
+    fn verify(&mut self, ctx: &mut HealthCheckContext) {
+        let validator = self.cluster.validator_instances();
+        let buffer = Duration::from_secs(30);
+        let timestamp = unix_timestamp_now() - buffer;
+        let commit_version =
+            match FullNodeHealthCheck::state_sync_committed_version(&self, timestamp, timestamp) {
+                Err(_e) => return,
+                Ok(s) => convert_index(s),
+            };
+        for (index, version) in &commit_version {
+            if version.len() < 2 {
+                panic!("Missed record to compare fullnode and validator version difference, record length: {}",
+                       version.len());
+            }
+            let mut version = version.clone();
+            version.sort();
+            // because validator version is the only negative number, it always be sorted at the first slot
+            for i in 1..version.len() {
+                // we define tolerance of version diff is 20000
+                if (version[i] + version[0]) > 20000_i64 || (version[i] + version[0]) < -20000_i64 {
+                    ctx.report_failure(
+                        validator[index.parse::<usize>().unwrap()]
+                            .clone()
+                            .peer_name()
+                            .clone(),
```

Comment:
I suspect one of these clones may be superfluous.

Reply:
aha, you are correct, just mis-added